lib/export_tables_to_big_query.rb

```diff
@@ -0,0 +1,254 @@
+        end
+      end
+      logging_details[:records_analyzed] += batch.size
+    end
+
+    logging_details = logging_details.merge({
+      records_missing_data: records_missing_data,
+      records_missing_data_count: records_missing_data.size
+    })
+    Rails.logger.info(logging_details.to_json)
+  end
+
+  def clean_record(record, table_name)
+    scratch = {}
+    record.as_json.keys.each { |key|
+      if record.class.columns_hash[key].type == :datetime || record.class.columns_hash[key].type == :date
+        # BigQuery uses the same standard format for dates and datetimes that `.to_s(:db)` outputs.
+        if record[key].blank?
+          scratch[key] = nil
+        else
+          scratch[key] = record[key].to_s(:db)
+        end
+      elsif key.match?(/data/) # Flatten the data structure into one level
+        if record.data.respond_to?(:keys)
+          data_field_normalizer[table_name].each do |normalized_key|
+            value = record.data[normalized_key]
+
+            # Dates are the only values in gias data that can be cast without running in to trouble with variations in
+            # the data. BigQuery already does this intelligently with every other column type. Miscast values are
+            # *sometimes* raised by BigQuery as errors (see next paragraph) and can be ignored if there are staistically
+            # insignificant numbers of them.
+            #
+            # Review the values on `data_Diocese_code` for a typical example - most are postcode-like strings, while a
+            # some are two or three digits numbers. Despite the presence of the two and three digit numbers, BigQuery
+            # sets the column type as sting and seems to be able to coerce the number-only values into strings. It does
+            # not do this reliably on every column of every table, however; it failed to coerce some larger numbers on
+            # the `School#address3` for example. That said, there are less than ten of these at the time of commit.
+            #
+            value = Date.parse(value).to_s(:db) if value.is_a?(String) && value.match?(/^\d{2}\-\d{2}\-\d{4}$/)
+            # I didn't use `#parameterize(separator: '_')` here because it is **SLOW** in contrast to this.
+            scratch["data_#{normalized_key.chomp(')').gsub(/\W+/, '_')}"] = value.presence
+          end
+        else
+          data_field_normalizer[table_name].each do |normalized_key|
+            scratch["data_#{normalized_key.chomp(')').gsub(/\W+/, '_')}"] = nil
+          end
+        end
+      elsif key == 'working_patterns'
+        # This attribute use the array_enum gem to map an array of integers to a text array. When the record is dumped
+        # using `as_json`, the gem does not do the conversion. Calling the non json version of the attribute ensures the
+        # data gets converted as expected.
+        scratch['working_patterns'] = record.working_patterns
+      elsif DROP_THESE_ATTRIBUTES.include?(key)
```

Comment:
Does it make sense to skip dropped attributes first?

Reply:
As per our conversation, we'll leave this for now.  Moving it is a lot of work and does little to mitigate the size or readability issues.